### PR TITLE
Fix vispy pixel offset in 3D multi-scale image layers

### PR DESCRIPTION
**Pull request by brisvag**
# References and relevant issues
See #6320 

# Description
This solves a specific problem in #6320 around 3D display of multi-scale image layers consistently with other data (e.g. 3D points data). As napari currently only ever shows the lowest resolution of a 3D image, we use the corresponding multi-scale offset to ensure that vispy starts drawing the 3D object at the intended location so that display is consistent between 2D and 3D, and with other layer types (e.g. `Points` is an easy one to test).

### CODE REVIEW

File: napari/_vispy/layers/base.py
Comment (andy-sweet) [inline napari/_vispy/layers/base.py]: Why do you subtract 1 from the last level's downsample factor? For a downsample factor of 1, we eventually get an offset of 0 whereas I'd expect 0.5 (UPDATE: rethought and simplified this).

Ideally, I'd prefer to handle this with the similar 2D array-like case below, which should work automatically due to the way [the tile2data transform is defined in slicing](https://github.com/napari/napari/blob/d02b8569270532cf9738b1bbc0ef90b926513b44/napari/layers/image/_slice.py#L219). But maybe I'm missing a detail here.

Comment (brisvag) [reply napari/_vispy/layers/base.py]: > For a downsample factor of 1, we eventually get an offset of 0 whereas I'd expect 0.5

No, because we don't want to offset the "base" scale. We do want the origin of the highest resolution to be at the center of the first pixel, to match every other thing in napari. If you leave the `-1` out, you end up shifting only multiscale layers so that they have the corner of the first pixel at the origin.

I hope this makes sense, I'm still jetlagged xD

> Ideally, I'd prefer to handle this with the similar 2D array-like case below

I tried to use the `tile2data` transform, but things broke down in way I could not figure out, and I was afraid of messing up unrelated things. If you know how to attack that, please do! In principle, I agree that this should be unified. On the other hand, there is a confusion/mixing of "multiscale" and "tiling" in this whole part of the code that makes it hard to work with, so it would be great if we could separate them more first.

**Review by Czaki — APPROVED** [review]

**Comment by Czaki** [comment]
@brisvag Should i mark this as 0.4.19?

File: napari/_vispy/layers/base.py
Comment (andy-sweet) [reply napari/_vispy/layers/base.py]: Will try to come back to this today or later in the week.

Agreed that these code paths are confusing. Maybe some motivation for @kephale to bring in multi-scale virtual data soon?

**Comment by brisvag** [comment]
If @andy-sweet approves, I would say yes, but he seemed to not like this approach!

**Comment by andy-sweet** [comment]
> If @andy-sweet approves, I would say yes, but he seemed to not like this approach!

I should be able to take a look at this within 48 hours, maybe 24 hours.

If I understand this bug correctly, I think it's been present for a long time, so I would vote to not include it in v0.4.19 if that's otherwise ready to go. But I don't feel too strongly.

File: napari/_vispy/layers/base.py
Comment (andy-sweet) [reply napari/_vispy/layers/base.py]: Thought some more on this and we shouldn't use `tile2data` here - that should only account for a partial image slice (e.g. in the 2D case) within the whole data, not for this extra offset which is for rendering.

Comment (andy-sweet) [inline napari/_vispy/layers/base.py]: `translate` has already been flipped to vispy axis ordering, whereas the downsample factors are still in numpy ordering, so you have to flip them here

```suggestion
            translate = translate + (self.layer.downsample_factors[-1][::-1] - 1) / 2
```

or define `translate` before flipping.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
The only blocking change is the following comment: https://github.com/napari/napari/pull/6319/files#r1362892161

An extra ask is to add a comment in the code to explain, *why* we handle 3D-multiscale images in this special way, as opposed to the proceeding special code block for 2D-images.

I tried #6320 before and after this change and agreed it fixes the rendering to match my expectation. But I don't understand all the interactions here.

I would feel a lot better with better test coverage, but that's not a requirement. I did start writing some tests, but not all of them pass (even after the changes here): https://gist.github.com/andy-sweet/546fb58f8be31ae67c2c40e8596de187

I kind of suspect that the code here doesn't work for all combos of multi-scale, layer offset/scale, and display dimensionality - either before or after these changes (so no need to verify/fix them all here either).

File: napari/_vispy/layers/base.py
Comment (andy-sweet) [reply napari/_vispy/layers/base.py]: Just to inspire confidence, I'm going to flip-flop again. But this time it's OK because I have a diagram!

napari-tile2data drawio

I'm going to claim that the `tile2data` transform captures part of information we care about here. The critical thing (which I missed and you alluded to) is that the `data` in `tile2data` is the data-space at level 0 (i.e. the highest resolution).

Slicing returns a chunk of image data. As with all image data we treat each pixel in that chunk as a sample. The `tile2data` transform tells us where those samples are in the data-space at level 0.

The other key constraint is that we want the sample at level 0 with the indices `(0, 0)` to be at the coordinate `(0.0, 0.0)` in that data space.

Therefore, in our current crude multi-scale approach, `tile2data` must take care of the offset and scale that take us from chunk sample indices to that level-0 data/index space.

For example, take the chunk index `(0, 1)`. According to the diagram, that needs to be transformed to `(0.5, 2.5)` in order to be consistent with the definitions above. This is achieved by a scale of `(2, 2)` and an offset of `(0.5, 0.5)`.

The other key factor here is that the vispy node texture needs another offset to account for the extent of the thing that it's rendering. That's what the code in `VispyBaseLayer._on_matrix_change` is trying to do.

Hope this helps. Will try to give this a bit more thought tomorrow.

**Review by andy-sweet — APPROVED** [review]
Approving to unblock this if we decide it should be brought into v0.4.19.

I think this bug has been around for several versions, so I would personally recommend against that since while this fix is a reasonable enough workaround, it adds to the existing confusion here.

**Comment by brisvag** [comment]
Ok, color me convinced, let's keep this out for now. Will change to draft and we can try to figure out a better approach after 0.4.19 along the lines of the comment above!

**Comment by jni** [comment]
Is this PR still relevant?

**Comment by psobolewskiPhD** [comment]
I think so -- see image.sc post from 3 days ago -- but we've not come to a good decision--nor has anyone else? e.g. https://github.com/ome/ngff/issues/89

**Comment by jni** [comment]
I think @andy-sweet's post [here](https://github.com/napari/napari/pull/6319#discussion_r1363143858) is a good starting point re decisions and matches the discussion in ome-ngff?

**Comment by brisvag** [comment]
Yeah I think we agreed that this PR has the "right solution" (it's more complicated for *different* layers with different scales). The doubt was about the implementation, and whether we're also breaking something else by doing this...

**Comment by andy-sweet** [comment]
> Yeah I think we agreed that this PR has the "right solution" (it's more complicated for _different_ layers with different scales). The doubt was about the implementation, and whether we're also breaking something else by doing this...

Agreed. The solution here seems right for the special case it fixes. That's the main reason I approved earlier - this fixes an issue on `main` and there isn't a better implementation yet.

I'm more open to including this in v0.5.0 than I was to including it in v0.4.19, because it feels like there a better opportunity to improve the implementation in later versions (i.e. v0.5.1, v0.5.2, etc.).

Before merging I think we should add an explanatory comment to this code, and I'll try to bring in one or two of the tests I wrote.

**Comment by brisvag** [comment]
Feel free to push to the branch @andy-sweet!

**Comment by andy-sweet** [comment]
> Feel free to push to the branch @andy-sweet!

Seems like I don't have permissions, so I made https://github.com/brisvag/napari/pull/31

**Comment by brisvag** [comment]
You should, you were able to merge main... Strange!

**Comment by andy-sweet** [comment]
> You should, you were able to merge main... Strange!

You're right, I was just being lazy when pushing.

**Comment by andy-sweet** [comment]
For reference, I also worked on a branch with a few more tests and a refactor of `_on_matrix_change` that helped me better understand what was going on (and maybe easier to understand/read?): https://github.com/napari/napari/compare/main...andy-sweet:napari:fix-vispy-offsets?expand=1

There were two remaining mysteries there.

1. For 2D we need to adjust the vispy node transform by (-0.5, -0.5), but for 3D we don't. In practice, it works, but I don't understand why. Does vispy treat these cases differently?

2. I think we're not handling `scale` correctly for the child offset related to overlays. That's a bug on that branch and on `main`. I think I know a solution on my branch, but would have to think a bit harder on `main` since I find the logic of `_on_matrix_change` harder to follow there.

**Comment by andy-sweet** [comment]
@brisvag : can I change the title and PR description here? This PR specifically fixes the issue for 3D (as described by the reproducer in #6320). Basic usage with 2D was already working. Though there may be some other related bugs for 2D and 3D!

**Comment by andy-sweet** [comment]
> @brisvag : can I change the title and PR description here? This PR specifically fixes the issue for 3D (as described by the reproducer in #6320). Basic usage with 2D was already working. Though there may be some other related bugs for 2D and 3D!

@brisvag : I updated the title and description to match my understand of this PR.

@Czaki or @jni : ideally one of you two would confirm this PR looks good, as I've made quite a few changes here now.

**Comment by brisvag** [comment]
@andy-sweet thanks, sorry I was away for some time! LGTM!

**Comment by jni** [comment]
Oof, I suspect this is not going to age well:

> The last downsample factor is used because we only ever show the
> last/lowest multi-scale level for 3D.

😂 but it does not bear worrying about right now. I'll merge! 😊 Thanks all!

**Comment by brisvag** [comment]
> I suspect this is not going to age well

AFAIK that will be easy to solve once the time comes: we just use the downsample factor relative to the current level :P

**Pull request merged**
